src/TimeSteppers/adams_bashforth.jl

```diff
@@ -75,89 +75,102 @@ end
 ##### Source term storage
 #####
 
+""" Store source terms for `u`, `v`, and `w`. """
+@kernel function ab2_store_velocity_source_terms!(G⁻, grid::AbstractGrid{FT}, χ, Gⁿ) where FT
+    i, j, k = @index(Global, NTuple)
+    @inbounds G⁻.u[i, j, k] = Gⁿ.u[i, j, k]
+    @inbounds G⁻.v[i, j, k] = Gⁿ.v[i, j, k]
+    @inbounds G⁻.w[i, j, k] = Gⁿ.w[i, j, k]
+end
+
+""" Store previous source terms for a tracer before updating them. """
+@kernel function ab2_store_tracer_source_term!(Gc⁻, grid::AbstractGrid{FT}, χ, Gcⁿ) where FT
+    i, j, k = @index(Global, NTuple)
+    @inbounds Gc⁻[i, j, k] = Gcⁿ[i, j, k]
+end
+
 """ Store previous source terms before updating them. """
 function ab2_store_source_terms!(G⁻, arch, grid, χ, Gⁿ)
 
-    # Velocity fields
-    @launch(device(arch), config=launch_config(grid, :xyz),
-            ab2_store_velocity_source_terms!(G⁻, grid, χ, Gⁿ))
+    workgroup, worksize = work_layout(grid, :xyz)
+
+    store_velocity_source_terms_kernel! = ab2_store_velocity_source_terms!(device(arch), workgroup, worksize)
+    store_tracer_source_term_kernel! = ab2_store_tracer_source_term!(device(arch), workgroup, worksize)
+
+    velocities_event = store_velocity_source_terms_kernel!(G⁻, grid, χ, Gⁿ)
+
+    events = [velocities_event]
 
     # Tracer fields
     for i in 4:length(G⁻)
         @inbounds Gc⁻ = G⁻[i]
         @inbounds Gcⁿ = Gⁿ[i]
-        @launch(device(arch), config=launch_config(grid, :xyz),
-                ab2_store_tracer_source_term!(Gc⁻, grid, χ, Gcⁿ))
+        tracer_event = store_tracer_source_term_kernel!(Gc⁻, grid, χ, Gcⁿ)
```

Comment:
missing a top event. If all these can run in parallel the event can be the same.